File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/MaximumDepthofBinaryTree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
        # @param root, a tree node
        # @return an integer
        def maxDepth(self, root):
        # either DFS and BFS, same performance, BFS need a queue
       
        '''     my original buggy code
         DFS
         need a variable to store max depth and curr depth
        def dfs(self, node, max_depth, curr_depth):
            if node = None:
                return
            curr_depth++;
            if( curr_depth > max_depth ):
                max_depth = curr_depth
            dfs(node.left)
            dfs(node.right) 
                return
        ''' 
            
            if root == None:
                return 0
            # self is an explicit argument refer to specific object
            # recursive way to get the max height of two children, very clean code
            return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/CPP/MaximumDepthofBinaryTree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int Depth( vector<TreeNode*> nodes, int height){
        vector<TreeNode*> newNodes;
    	for(auto _it=std::begin(nodes); _it!=end(nodes); _it++){
		//for(vector<TreeNode*>::iterator _it=nodes.begin(); _it!=nodes.end(); _it++){
            if( (*_it)->left!=NULL)
                newNodes.push_back( (*_it)->left);
            if(  (*_it)->right!=NULL)
                newNodes.push_back(  (*_it)->right);
        }
        
        if( newNodes.size() == 0)
            return height;
        else
            return Depth(newNodes, ++height);
    }
    
    int maxDepth(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        
        if(root!=NULL){
            vector<TreeNode*> nodes;
            nodes.push_back(root);
            return Depth(nodes,1);
        }
        else
            return 0;
    }
};
